### ISSUE TITLE
reverse モードの勾配が ts[0] へのcotangentを反映しない

### ISSUE THREAD

**Issue by kshkss** [kshkss/daex#2]
## 現象

`daeint(..., mode="reverse")`（デフォルト）を `jax.grad` で微分すると、損失関数が最初の出力点（`ts[0]`、すなわち渡した初期値 `y0` と同一の値）に依存する場合、その依存分のcotangentが勾配に反映されない。

## 再現方法

```python
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp
from typing import NamedTuple
from daex.semi_explicit import daeint, def_semi_explicit_dae

class State(NamedTuple):
    x: jax.Array
    y: jax.Array

class Params(NamedTuple):
    a: jax.Array

def derivative(params, t, stat):
    return State(x=None, y=stat.x * params.a * t)

def constraint(params, t, stat):
    return stat.x ** 2 - jnp.sum(stat.y)

params = Params(a=jnp.array(1.0))
y0 = State(y=jnp.array(2.0), x=jnp.sqrt(2.0))
dae = def_semi_explicit_dae(derivative, constraint, params, jnp.array(0.0), y0)
ts = jnp.linspace(-0.5, 1.0, 11)
opts = dict(quad_order=9, options_adj={"calc_initcond": "yp0", "calc_init_dt": -0.01})

def loss(params, ts, y0):
    u, _ = daeint(params, dae, ts, y0, mode="reverse", **opts)
    return jnp.sum(u.y)  # ts[0] の項を含む

grad = jax.grad(loss, argnums=2)(params, ts, y0)
print("jax.grad:", grad.y)          # => 10.0729...

eps = 1e-6
y0p = State(x=y0.x, y=y0.y + eps)
y0m = State(x=y0.x, y=y0.y - eps)
fd = (loss(params, ts, y0p) - loss(params, ts, y0m)) / (2 * eps)
print("finite diff:", fd)           # => 11.0731...  (ちょうど cotangent[0]=1.0 分ズレる)
```

損失を `ts[0]` を含まない形（例: `u.y[5]` や `u.y[-1]` のみ）にすると `jax.grad` と有限差分は一致する。

## 原因

`_daeint_bwd2`（`src/daex/semi_explicit.py`、`_daeint2` の `custom_vjp` の `bwd` 側）内の `fori_loop` は、反転済みの cotangent 配列に対して添字 `0` から `points - 2` までしか処理しない。反転前の添字に戻すと、これは元の配列の添字 `points-1` から `1` までに相当し、**添字 `0`（`ts[0]`）の cotangent は一度もループ本体に渡されない**。

```python
dJda, dJdt0_prev, dJdt, dJdy0 = jax.lax.fori_loop(
    0, points - 1, body,
    (jnp.zeros_like(params), 0.0, jnp.zeros(points), jnp.zeros_like(wy[0])),
)
...
return (dJda, dJdt[::-1], None, dJdy0)
```

`_daeint2` の出力の1行目（`ts[0]` に対応する `x`, `y`）は、渡された初期値 `x0`, `y0` そのものである（恒等写像）。そのため、損失がこの出力点に依存する場合、`d(loss)/d(y0)` にはそのcotangent（`wy[0]`）がそのまま直接加算されるべきだが、`_daeint_bwd2` はこれを一切加算せずに `dJdy0` を返している。

この問題は、新規追加した `adjoint()` 関数（随伴軌道 z(t) を返す公開API、各点のcotangentを正しく累積する実装）の検証中に見つかった。`adjoint()` の `z(ts[0])` は有限差分と一致するが、既存の `jax.grad` の結果はそれと `wy[0]` 分だけズレる。

## 影響範囲

`daeint(..., mode="reverse")`（デフォルト）の勾配計算全般。損失関数が出力配列の最初の点（`ts[0]`）に依存する場合にのみ影響する。損失が `ts[0]` を含む中間点・終端点のみに依存する場合は問題ない。

`mode="forward"` は無関係（別の実装経路のため影響なし）。

## 修正方針（案）

`_daeint_bwd2` の `fori_loop` 終了後、返す `dJdy0` に元の配列の添字 `0` の cotangent（`wy[0]`、反転前）を直接加算する。

## 受け入れ条件

- 損失が `ts[0]` に依存する場合の `jax.grad(daeint(...))` の勾配が、有限差分と許容誤差内で一致する。
- 既存の回帰テスト（`tests/test_semi_explicit_modes.py`, `tests/test_adjoint.py`）が全てpassし続ける。
- `ts[0]` に依存する損失に対する回帰テストを追加する。
